plumbing/format/packfile/delta_selector.go

```diff
@@ -0,0 +1,176 @@
+	maxDepth = int64(50)
+)
+
+// set of object types that we should apply deltas
+var applyDelta = map[plumbing.ObjectType]bool{
+	plumbing.BlobObject: true,
+	plumbing.TreeObject: true,
+}
+
+type deltaSelector struct {
+	storer storer.ObjectStorer
+	depths map[plumbing.Hash]int
+}
+
+func newDeltaSelector(s storer.ObjectStorer) *deltaSelector {
+	return &deltaSelector{
+		storer: s,
+	}
+}
+
+func (dw *deltaSelector) getObjectsToPack(hashes []plumbing.Hash) ([]*plumbing.ObjectToPack, error) {
+	var objectsToPack []*plumbing.ObjectToPack
+	for _, h := range hashes {
+		o, err := dw.storer.Object(plumbing.AnyObject, h)
+		if err != nil {
+			return nil, err
+		}
+
+		objectsToPack = append(objectsToPack, plumbing.NewObjectToPack(o))
+	}
+
+	return objectsToPack, nil
+}
+
+// GetObjectsToPack creates a list of ObjectToPack from the hashes provided,
+// creating deltas if it's suitable, using an specific internal logic
+func (dw *deltaSelector) GetObjectsToPack(hashes []plumbing.Hash) ([]*plumbing.ObjectToPack, error) {
+	otp, err := dw.getObjectsToPack(hashes)
+	if err != nil {
+		return nil, err
+	}
+
+	dw.sort(otp)
+
+	if err := dw.walk(otp); err != nil {
+		return nil, err
+	}
+
+	return otp, nil
+}
+
+func (dw *deltaSelector) sort(objectsToPack []*plumbing.ObjectToPack) {
+	sort.Sort(byTypeAndSize(objectsToPack))
+}
+
+func (dw *deltaSelector) walk(objectsToPack []*plumbing.ObjectToPack) error {
+	for i := 0; i < len(objectsToPack); i++ {
+		target := objectsToPack[i]
+
+		// We only want to create deltas from specific types
+		if !applyDelta[target.Original.Type()] {
+			continue
+		}
+
+		for j := i - 1; j > 0; j-- {
+			base := objectsToPack[j]
+			// Objects must use only the same type as their delta base.
+			if base.Original.Type() != target.Original.Type() {
+				break
+			}
+
+			delta, err := dw.tryToDeltify(base, target)
+			if err != nil {
+				return err
+			}
+
+			if delta != nil {
+				objectsToPack[i] = delta
+			}
+		}
+	}
+
+	return nil
+}
+
+func (dw *deltaSelector) tryToDeltify(
+	base, target *plumbing.ObjectToPack) (*plumbing.ObjectToPack, error) {
+
+	// If the sizes are radically different, this is a bad pairing.
+	if target.Original.Size() < base.Original.Size()>>4 {
+		return nil, nil
+	}
+
+	msz := dw.deltaSizeLimit(base, target)
+	// Nearly impossible to fit useful delta.
+	if msz <= 8 {
+		return nil, nil
+	}
+
+	// If we have to insert a lot to make this work, find another.
+	if target.Original.Size()-base.Original.Size() > msz {
+		return nil, nil
+	}
+
+	// Now we can generate the delta using originals
+	delta, err := GetOFSDelta(base.Original, target.Original)
+	if err != nil {
+		return nil, err
+	}
+
+	deltaToPack := plumbing.NewDeltaObjectToPack(base, target.Original, delta)
+
+	// if delta better than target
+	if target.IsDelta() {
+		newMsz := dw.deltaSizeLimit(base, deltaToPack)
+		if newMsz > msz {
+			return target, nil
+		}
+	}
+
+	return deltaToPack, nil
```

Comment:
instead of returning a new ObjectToPack, you could just set the base, delta and depth. Maybe something like:
target.SetBase(*ObjectToPack)